service/database.go

```diff
@@ -0,0 +1,51 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2016-2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package service
+
+import (
+	"database/sql"
+	"log"
+
+	// postgresql driver
+	_ "github.com/lib/pq"
+)
+
+// Datastore interface for the database logic
+type Datastore interface {
+	ListModels() ([]Model, error)
+	CreateModelTable() error
+}
+
+// DB local database interface with our custom methods.
+type DB struct {
+	*sql.DB
+}
+
+// OpenSysDatabase Return an open database connection
+func OpenSysDatabase(driver, dataSource string) *DB {
+	// Open the database connection
+	db, err := sql.Open(driver, dataSource)
+	if err != nil {
+		log.Fatal(err)
+	} else {
+		log.Println("Database opened successfully.")
```

Comment:
sql.Open does not actually opens the database, so this will succeed even if there is no database available, and then it will fail elsewhere when trying to access it (when accessing the models endpoint)

"Perhaps counter-intuitively, sql.Open() does not establish any connections to the database, nor does it validate driver connection parameters. Instead, it simply prepares the database abstraction for later use. The first actual connection to the underlying datastore will be established lazily, when it’s needed for the first time. If you want to check right away that the database is available and accessible (for example, check that you can establish a network connection and log in), use db.Ping() to do that, and remember to check for errors:

err = db.Ping()
if err != nil {
	// do something here
}
"
(As explained at: http://go-database-sql.org/accessing.html)